Hud Image Culling By Z Level: Theft edition

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
I'm re-pring kyler's initial work (#63558), since he's busy at the moment, and I think it would be useful.
I've had to make a minor change to how alternate appearance huds manage their key (314e5f70175a860b45e182c1c42852ced3850523)
Not 100% sure if this is the right way to go about it, the type is honestly kinda a deadend, but it seemed sane. Feedback'd be nice.

The rest of this is kyler's, cause he's based.

makes hud images only show to players on the same z level and linked z levels above to optimize load on sendmaps a bit more. also cleans up the code a bit. eventually i want to make huds support currently inactive images to improve savings more but ill do that later
fixes https://github.com/tgstation/tgstation/issues/63872
fixes https://github.com/tgstation/tgstation/issues/63901

## Why It's Good For The Game


so theres this step of the SendMaps profiler called "Check HUD/image visual contents". each tick, SendMaps iterates through every clients client.screen and client.images and checks each one for vis_contents and looking through a lot of sendmaps profiles it seems to take around 10% of all sendmaps varying a fair bit. for some reason images take about a fifth as long as screen movables to scan in this step buuuut a few days ago i realized that every ghost seems to have
![image](https://user-images.githubusercontent.com/58055496/155864438-6f778c85-7de6-4a35-8708-8a32f4954067.png)
images. and all of this it seems like is from atom_hud datums. it turns out that every single ghost gets: around 2 images of every mob in the world, 1 image of every door in the world, 1 image of every airlock in the world, 1 image of every bot path in the world, among presumably a few other painful examples. this is because they spawn with the advanced medical, security, and worst of all diagnostic huds. so every atom in the world those huds apply to are added to every ghost clients images list and increase maptick. each ghost also has around 60 screen movables from plane masters and such, but 60 * 5 is just 300, so screen objects are getting outweighed here by an order of magnitude. luckily most of the time ghosts are a minority of players, but on long rounds with lots of dead players i suspect the higher amounts of them can push maptick per person above 0.5

so by restricting it to images only in the players z level i (on local so the raw image count is lower than itd be on live) lowered it to around 300 on tramstation upper and lower levels and 500 on lavaland (before it was about 1600 on local)

Kylerace committed:
Merge remote-tracking branch 'upstream/master'

Kylerace committed:
makes hud images only apply by z level

Kylerace committed:
makes some of the atom_hud procs have better names

Kylerace committed:
fixes warning with the hud_user list and adds better documentation

Kylerace committed:
better docs for hud_images

Kylerace committed:
removes TODOs

Kylerace committed:
docs for hud_list

Kylerace committed:
adds support for linked z levels so mobs can see lower ones

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into hud_images_by_z_level

Kylerace committed:
fixes merge conflict and shittily makes only shocked airlocks get added

Kylerace committed:
adds support for setting images in the hud as active and inactive

Kylerace committed:
gets rid of unatomic spatial grid change

Kylerace committed:
Merge remote-tracking branch 'origin/master' into hud_images_by_z_level

Kylerace committed:
maybe i should actually try COMPILING my changes

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into hud_images_by_z_level

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into hud_images_by_z_level

Kylerace committed:
fixes merge skew and makes it compile again

Kylerace committed:
fixes huds refusing to remove from users who changed z level

Kylerace committed:
improves z level and registration logic

Kylerace committed:
fixes antag huds not appearing

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into test-aced

LemonInTheDark committed:
Fixes antag huds not properly setting. We now use hud_list in init, so it needs to be set before the new call, not after. Not sure why the use of appearance key was split like this, but none else knows either so none can stop me

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into test-aced

LemonInTheDark committed:
Ensures that hiding a basic appearance also hides the atom's active list too

LemonInTheDark committed:
Fixes antag huds going poof

Ensures that remove_atom_from_hud will return false if the passed atom
isn't managed by it

This fixes antag huds disappearing randomly, since they assumed that if
the parent call of remove_atom_from_hud returned true, we should delete
ourselves. This is a safe assumption for them to make, since they should
only ever have one atom.

Does kinda bork if we call remove_atom_from_hud in a way that is unsure
if the passed atom is actually in that list. We were forced into doing
this by how atom huds use the qdeleting signal.

LemonInTheDark commented:
Huds aren't constructed in the way kyler initially thought, which makes parts of this harder. will get back to you

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into test-aced

realthomas69 commented:
this appears to have fucked cult action buttons, it only shows two at once

LemonInTheDark commented:
What? This should not be testmerged

Kylerace committed:
makes basic alternate_appearance's only update themselves when setting their hud image to active and makes them not add themselves to the global huds_by_category list

Kylerace committed:
fixes mistake with hud_users list being set non associatively (bad)

Kylerace committed:
as anything in bot path loops

LemonInTheDark committed:
(hopefully) fixes antag huds

basically just makes basic alternate appearances 1. not add themselves to the global huds_by_category list since they dont work with it, and 2. adds an option in /atom/proc/make_hud_image_active() to only update a single atom hud, which basic alternate appearances do.

team antag huds seem to work fine. combo huds dont work question mark (might just be how im testing them on local)
edit: nm combo huds seem to work fine they just only apply in weird situations both on master and here

Why It's Good For The Pull Request

it makes it better

LemonInTheDark commented:
lmao. Will fix that tomorrow, kyler put something in place to deal with non global huds, so we should be good to retm this once I clean things up

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into test-aced

LemonInTheDark committed:
Fixes merge skew problems

LemonInTheDark committed:
Merge branch 'test-aced' of https://github.com/LemonInTheDark/tgstation into test-aced

LemonInTheDark committed:
Makes bot paths non global

This way they can show themselves to only the bot that "owns" them, ya
feel me?

iprice commented:
Ghost/admin HUDs not showing player job on Terry presently (shows the generic ? ), probably related to this

LemonInTheDark commented:
Likely yes. I will look into it, thanks a ton.

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into test-aced

LemonInTheDark committed:
Fixes huds not showing up sometimes, cleans up some code

Post Kapu's limb refactor, we were calling prepare_huds twice in a human
init call chain. What was happening was this:

call prepare_huds() // Human
I gained a new hud image
I set active hud icons to mirror it
call prepare_huds() // Living
I overwrote the new hud image
I attempted to set active hud icons, which failed because it assumes
this can never happen

*cries*

LemonInTheDark commented:
> Ghost/admin HUDs not showing player job on Terry presently (shows the generic ? ), probably related to this

Handled, damn you kapu

TheBonded commented:
both of the linked issue reports are already closed, but the comments seem to suggest they arent fixed?

LemonInTheDark commented:
They were testmerge bugs, and I am dumb

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into test-aced

LemonInTheDark committed:
Renames add_hud_to_atom to show_to

My hope is this will make understanding hud code a bit easier, by tying
the behavior to a "verb" more closely. Also renamed a few vars

LemonInTheDark committed:
remove_hud_from_mob -> hide_from

LemonInTheDark committed:
Nitpicks a few comments

LemonInTheDark committed:
Whoops/fuck/shit/damn it all/hhhhhhhhhhhh

Kylerace commented on code/datums/hud.dm:
```diff
@@ -17,116 +21,356 @@ GLOBAL_LIST_INIT(huds, list(
-	if(!A)
+///returns a list of all hud users in the given z level and linked upper z levels (because hud users in higher z levels can see below)
+/datum/atom_hud/proc/get_hud_users_for_z_level(z_level)
+	if(z_level <= 0)
```
should combine this with the stack_trace() below

LemonInTheDark replied:
yeet

LemonInTheDark committed:
Moves check down, improves stack trace a bit
